src/lib/esnext.object.d.ts

```diff
@@ -0,0 +1,11 @@
+interface ObjectConstructor {
+    /**
+     * Groups members of an iterable according to the return value of the passed callback.
+     * @param items An iterable.
+     * @param keySelector A callback which will be invoked for each item in items.
+     */
+    groupBy<K extends PropertyKey, T>(
+        items: Iterable<T>,
+        keySelector: (item: T, index: number) => K,
+    ): Partial<Record<K, T[]>>;

```

Comment:
why is the return type `Partial`?

Reply:
Because [you might not actually get all of the keys](https://github.com/microsoft/TypeScript/issues/47171#issuecomment-1848451195).

As a trivial example, consider the case that the iterable is empty. Then the resulting record will have no keys at all. Typing as `Record<K, T[]>` [would mean that](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgPICMBWEFgMID2IAzmFAK64FTIDeAUMk8gOZQHkAOAQgJ4A8AaWQQAHpBAATYsgAK7TtDC9BEXgBpkAFQB8ACkbMjwSAFtiALmQBJSFDjoANhH671ho0wDWagMoRnKigrPRMIUystTVBJMSsQclN0aABKZABeHWRBdyMUqwAlHGpJIU0tAG0AXR0AbnoAX3p6BCJSZHQ4YmAEDLQsHDAAOjYOHl49CoAOKs1RDKz5-mQAVmQAfmQAcmJTOEdHLeQrLcc4KBYILZT6+gB6O+QAKheX5DAAC2AZYg+OR0kHRQcBAIig7CgAEJ6J1ugghrt9o4hs4QCxPvUgA) e.g. `Object.groupBy(vals, x => x < 5 ? 'small' : 'large')` would be typed as _always_ having `small` and `large` keys, such that `result.small.length` would check without errors, which is misleading.

Reply:
Ah, right, I forgot that `Map<K, T[]>` has `get: (k: K) => T[] | undefined` so was confused by the difference in types.